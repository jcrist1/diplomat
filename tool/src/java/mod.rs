use std::collections::HashSet;
use std::fs::File;
use std::io::Write;
use std::path::Path;

use diplomat_core::hir::TypeContext;

use crate::c2;
use crate::common::FileMap;

const TMP_C_DIR: &str = "tmp";
const LIBRARY: &str = "somelib"; // todo: build from conf. Ensure that name is not the same as any
                                 // type
const GROUP: &str = "dev.diplomattest"; // todo: config
const TMP_LIB_NAME: &str = "dev/diplomattest/somelib"; // todo: build from conf
const JAVA_DIR: &str = "src/main/java/";

pub fn run(
    tcx: &TypeContext,
    _conf_path: Option<&Path>,
    out_folder: &Path,
) -> std::io::Result<FileMap> {
    let files = FileMap::default();
    let mut context = c2::CContext::new(tcx, files, false);
    context.run();

    let errors = context.errors.take_all();

    if !errors.is_empty() {
        eprintln!("Found errors when generating c  code");
        for error in errors {
            eprintln!("\t{}: {}", error.0, error.1);
        }
    }

    let out_files = context.files.take_files();

    let tmp_path = out_folder.join(TMP_C_DIR);
    std::fs::create_dir(&tmp_path)?;
    let mut include_files = HashSet::new();
    for (subpath, text) in out_files {
        let out_path = tmp_path.join(&subpath);
        if !subpath.ends_with(".d.h") && subpath.ends_with(".h") {
            include_files.insert(subpath);
        }
        let parent = out_path
            .parent()
            .expect("Cannot create files at top level dir /");
        std::fs::create_dir_all(parent)?;
        let mut out_file = File::create(&out_path)?;
        out_file.write_all(text.as_bytes())?;
    }

    let lib_path = tmp_path.join(format!("{LIBRARY}.h"));

    let mut lib_file = File::create(&lib_path)?;
    for include in include_files {
        writeln!(lib_file, "#include \"{include}\"")?;
    }

    // jextract \
    //   --include-dir /path/to/mylib/include \
    //   --output src \
    //   --target-package org.jextract.mylib \
    //   --library mylib \
    //   /path/to/mylib/include/mylib.h

    let package = format!("{GROUP}.{LIBRARY}.ntv");
    let mut command = std::process::Command::new("jextract");
    command
        .arg("--include-dir")
        .arg(&tmp_path)
        .arg("--output")
        .arg(out_folder)
        .arg("--target-package")
        .arg(package)
        .arg("--library")
        .arg(LIBRARY)
        .arg(lib_path);

    println!("Running: {:?}", command);

    // todo: delete directory

    match command.output() {
        Err(err) => match err.kind() {
            std::io::ErrorKind::NotFound => {
                eprintln!("Check that jextract is in your path and all directories exist. See https://github.com/openjdk/jextract/blob/5715737be0a1a9de24cce3ee7190881cfc8b1350/doc/GUIDE.md");
                return Err(err);
            }
            _ => return Err(err),
        },
        Ok(ok) => {
            let stdout = String::from_utf8_lossy(&ok.stdout);
            println!("Output from jextract:\n{stdout}");

            let stderr = String::from_utf8_lossy(&ok.stderr);
            println!("Std Err from jextract:\n{stderr}");
        }
    }

    let files = FileMap::default();
    Ok(files)
}
