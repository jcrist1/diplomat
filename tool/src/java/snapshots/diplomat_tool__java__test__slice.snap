---
source: tool/src/java/mod.rs
expression: res
---
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;

public class Float64Vec {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    static class Float64VecCleaner implements Runnable {

        MemorySegment segment;
        Float64VecCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Float64Vec_destroy(this.segment);
        }
    }

    Float64Vec() {}
    
    public static Float64Vec new_(double[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vMemSeg = arena.allocateFrom(JAVA_DOUBLE, v);
            var nativeVal = somelib_h.Float64Vec_new(vMemSeg, vLen);
            var returnVal = new Float64Vec();
            returnVal.internal = nativeVal;
            var cleaner = new Float64Vec.Float64VecCleaner(nativeVal);
            returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    
    public double[] asBoxedSlice() {
        
        var boxArena = Arena.ofConfined();
        
        var nativeVal = somelib_h.Float64Vec_as_boxed_slice(boxArena, internal);
        var data = dev.diplomattest.somelib.ntv.DiplomatF64View.data(nativeVal);
        var len = dev.diplomattest.somelib.ntv.DiplomatF64View.len(nativeVal);
        return SliceUtils.doubleSliceToArray(nativeVal);
    }
    
    public double[] asSlice() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.Float64Vec_as_slice(arena, internal);
            var data = dev.diplomattest.somelib.ntv.DiplomatF64View.data(nativeVal);
            var len = dev.diplomattest.somelib.ntv.DiplomatF64View.len(nativeVal);
            return SliceUtils.doubleSliceToArray(nativeVal);
        }
    }
    
}
============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;

public class MyString {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    static class MyStringCleaner implements Runnable {

        MemorySegment segment;
        MyStringCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.MyString_destroy(this.segment);
        }
    }

    MyString() {}
    
    public MyString(String v) {
        
        try (var arena = Arena.ofConfined()) {
            var vMemSeg = arena.allocateFrom(v, StandardCharsets.UTF_8);
            var vLen = vMemSeg.byteSize();
            var nativeVal = somelib_h.MyString_new(vMemSeg, vLen - 1);
            this.internal = nativeVal;
            var cleaner = new MyString.MyStringCleaner(nativeVal);
            this.cleanable = Lib.cleaner.register(this, cleaner);
        }
    }
    
    public static MyString newUnsafe(String v) {
        
        try (var arena = Arena.ofConfined()) {
            var vMemSeg = arena.allocateFrom(v, StandardCharsets.UTF_8);
            var vLen = vMemSeg.byteSize();
            var nativeVal = somelib_h.MyString_new_unsafe(vMemSeg, vLen - 1);
            var returnVal = new MyString();
            returnVal.internal = nativeVal;
            var cleaner = new MyString.MyStringCleaner(nativeVal);
            returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    public static MyString newOwned(String v) {
        
        var vMemSeg = Arena.global().allocateFrom(v, StandardCharsets.UTF_8);
        var vLen = vMemSeg.byteSize();
        var nativeVal = somelib_h.MyString_new_owned(vMemSeg, vLen - 1);
        var returnVal = new MyString();
        returnVal.internal = nativeVal;
        var cleaner = new MyString.MyStringCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static MyString newFromFirst(String [] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vData = SliceUtils.strs8(arena, v);
            var vLen = v.length;
            var nativeVal = somelib_h.MyString_new_from_first(vData, vLen);
            var returnVal = new MyString();
            returnVal.internal = nativeVal;
            var cleaner = new MyString.MyStringCleaner(nativeVal);
            returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    
    public void setStr(String newStr) {
        
        try (var arena = Arena.ofConfined()) {
            
            var newStrMemSeg = arena.allocateFrom(newStr, StandardCharsets.UTF_8);
            var newStrLen = newStrMemSeg.byteSize();
            somelib_h.MyString_set_str(internal, newStrMemSeg, newStrLen - 1);
            
        }
    }
    
    public String getStr() {
        
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.MyString_get_str(internal, writeable);
        var buffer = DiplomatWrite.buf(writeable);
        var string = buffer.getString(0, StandardCharsets.UTF_8);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
    public String getBoxedStr() {
        
        var boxArena = Arena.ofConfined();
        
        var nativeVal = somelib_h.MyString_get_boxed_str(boxArena, internal);
        return SliceUtils.readUtf8(nativeVal);
    }
    
}
============================
